File: src/java/EntityFX/src/EntityFX/Core/Dhrystone/ParallelDhrystoneBenchmark.java
package EntityFX.Core.Dhrystone;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import EntityFX.Core.Writer;
import EntityFX.Core.Generic.BenchResult;
import EntityFX.Core.Generic.BenchmarkBase;

public class ParallelDhrystoneBenchmark extends BenchmarkBase<BenchResult[]> {

    public ParallelDhrystoneBenchmark(Writer writer, boolean printToConsole) throws FileNotFoundException {
        super(writer, printToConsole);
        Ratio = 4;
        isParallel = true;
    }

    @Override
    public BenchResult[] benchImplementation() throws InterruptedException, ExecutionException {
        return benchInParallel(() -> {
            try {
                return new Dhrystone2(false);
            } catch (FileNotFoundException e1) {
                e1.printStackTrace();
            }
            return null;
        }, a -> {
            try {
                return a.Bench(Dhrystone2.LOOPS);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }, (a, r) -> {
            r.Points = a.VaxMips * Ratio;
            r.Result = a.VaxMips;
            r.Output = a.Output;
        });
    }

    @Override
    public BenchResult populateResult(BenchResult benchResult, BenchResult[] dhrystonBenchResults) {
        var result = buildParallelResult(benchResult, dhrystonBenchResults);
        result.Result = Arrays.stream(dhrystonBenchResults).mapToDouble(r -> r.Result).sum();
        result.Units = "DMIPS";
        result.Output = String.join("", Arrays.toString(Arrays.stream(dhrystonBenchResults).map(r -> r.Output).toArray()));
        return result;
    }

    @Override
    public void warmup(Double aspect) {
    }
}

File: src/python/entityfx/parallel_dhrystone_benchmark.py
from entityfx.dhrystone2 import Dhrystone2
from entityfx.benchmark_base import BenchmarkBase
from entityfx.writer import Writer

class ParallelDhrystoneBenchmark(BenchmarkBase):

    def __init__(self, writer: Writer, print_to_console : bool=True) -> None:
        super().__init__(writer, print_to_console)
        self.is_parallel = True
        self.ratio = 4
    
    def benchImplementation(self) -> list:
        return self.bench_in_parallel(lambda : Dhrystone2(False), lambda a: a.bench(), lambda a, r : self.setBenchResultFunc(a, r))

    def warmup(self, aspect : float=.05) -> None:
        pass

    def populateResult(self, bench_result, dhrystone_result : list):
        result = self._buildParallelResult(bench_result, dhrystone_result) 
        result["Result"] = sum(map(lambda x : x["Result"], dhrystone_result))
        result["Units"] = "DMIPS"
        result["Output"] = "".join(map(lambda x : x["Output"], dhrystone_result))
        return result


    def setBenchResultFunc(self, a, r):
        r["Points"] = a["VaxMips"] * self.ratio
        r["Result"] = a["VaxMips"]
        r["Output"] = a["Output"]